### DISCUSSION TITLE
Duplicate array entries when multiple inputs are specified

### DISCUSSION THREAD

**Discussion by keithpine** [mikefarah/yq#1219] [Q&A]
Not sure if this is a bug, or a misunderstanding on my part about how multiple inputs work. I distilled my problem into this minimal example:

```text
❯ cat a.yml b.yml c.yml
{}
{}
{}

❯ yq eval-all '(select(fi==0) | .z) += [1] | select(fi==0)' a.yml
z:
  - 1

❯ yq eval-all '(select(fi==0) | .z) += [1] | select(fi==0)' a.yml b.yml
z:
  - 1
  - 1

❯ yq eval-all '(select(fi==0) | .z) += [1] | select(fi==0)' a.yml b.yml c.yml
z:
  - 1
  - 1
  - 1
```

You can see that for every additional input file that is given, this expression results in a duplicate array entry (multiplied). I would only expect one entry in the end, no matter how many files are input.

My original use case is to copy data from one YAML file and into another with a different structure. I thought it had to do with using references, but it really just seems like it just depends on the number of files specified as input.

```text
❯ cat to.yml from.yml
{}
x: 1

❯ yq eval-all -i '
select(fi==0) ref $to | select(fi==1) ref $from
| $to.a.b += [$from.x]
| $to' to.yml from.yml

❯ cat to.yml
a:
  b:
    - 1
    - 1
```
In this case I'd just expect `a.b = [1]`.

**Comment by keithpine** [comment]
I guess that's how `eval-all` works, it's going to run the same expression on each of the inputs. It's more clear to me when I see it like this:

```text
❯ yq eval-all '(select(fi==0) | .x) = [1]' a.yml b.yml
x:
  - 1
  - 1
---
{}
```

Is there a way to achieve what I'm trying to do?

**Reply by keithpine** [reply]
Actually, I'm still confused on this example. If the expression is just an array assignment, why does it end up appending?

**Comment by mikefarah** [comment]
This looks like a bug related to https://github.com/mikefarah/yq/issues/1214 - looking into it

**Reply by mikefarah** [reply]
Sorry - I meant: https://github.com/mikefarah/yq/issues/1206

**Comment by mikefarah** [comment]
Ok I can see what's happening here. `eval-all` loads each file into context, then on the RHS expression the `[1]` is evaluated for each item in context, which yields `[1]` for the first file, and again the second. This seems odd - but I (think) it makes sense as its consistent with how everything else works. Going to keep thinking about it - till then you can either:

a) not use `eval-all` (eval all is only really required if you want to execute an expression against all files simultaneously, like merging file 1 into file 2 or something. The default `eval` will get what you want, at least in the example your provided.

b) use the 'with' operator to restrict both the LHS and RHS sides: `with(select(fi==0); .x += [1])`
